common/src/attestations/signer.rs

```diff
@@ -5,48 +5,118 @@ use alloy_primitives::Address;
 use eip_712_derive::{
     sign_typed, Bytes32, DomainSeparator, Eip712Domain, MemberVisitor, StructType,
 };
-use ethers::utils::hex;
+use ethers::{
+    signers::{coins_bip39::English, MnemonicBuilder, Signer, Wallet},
+    utils::hex,
+};
 use ethers_core::k256::ecdsa::SigningKey;
 use ethers_core::types::U256;
 use keccak_hash::keccak;
 use secp256k1::SecretKey;
 use std::convert::TryInto;
 use toolshed::thegraph::DeploymentId;
 
+use crate::prelude::Allocation;
+
+pub fn derive_key_pair(
+    indexer_mnemonic: &str,
+    epoch: u64,
+    deployment: &DeploymentId,
+    index: u64,
+) -> Result<Wallet<SigningKey>, anyhow::Error> {
+    let mut derivation_path = format!("m/{}/", epoch);
+    derivation_path.push_str(
+        &deployment
+            .to_string()
+            .as_bytes()
+            .iter()
+            .map(|char| char.to_string())
+            .collect::<Vec<String>>()
+            .join("/"),
+    );
+    derivation_path.push_str(format!("/{}", index).as_str());
+
+    Ok(MnemonicBuilder::<English>::default()
+        .derivation_path(&derivation_path)
+        .expect("Valid derivation path")
+        .phrase(indexer_mnemonic)
+        .build()?)
+}
+
+pub fn attestation_signer_for_allocation(
+    indexer_mnemonic: &str,
+    allocation: &Allocation,
+) -> Result<SigningKey, anyhow::Error> {
+    // Guess the allocation index by enumerating all indexes in the
+    // range [0, 100] and checking for a match
+    for i in 0..100 {
+        // We try created_at_epoch and created_at_epoch-1 here for the following reason:
+        //
+        // Let's say the indexer has prepared an allocation transaction while the
+        // current epoch was N . By the time the transaction actually makes its into
+        // the blockchain, the epoch may still be N, or it may be N+1.
+        //
+        // If the transaction was mined during epoch N, then `created_at_epoch` will be N
+        // and the allocation ID will match that. If the transaction was mined durign epoch
+        // N+1, then `created_at_epoch` will be N+1 but the allocation ID will have been
+        // created using `created_at_epoch-1 = N`.
+        for created_at_epoch in [allocation.created_at_epoch, allocation.created_at_epoch - 1] {
+            let allocation_wallet = derive_key_pair(
+                indexer_mnemonic,
+                created_at_epoch,
+                &allocation.subgraph_deployment.id,
+                i,
+            )?;
+            if allocation_wallet.address().as_fixed_bytes() == allocation.id {
+                return Ok(allocation_wallet.signer().clone());
+            }
+        }
+    }
+    Err(anyhow::anyhow!(
+        "Could not find allocation signer for allocation {}",
+        allocation.id
+    ))
+}
```

Comment:
This can be refactored to use `wallet_for_allocation()`

Reply:
It's actually unused now. I've updated the tests for it to be tests for `AttestationSigner::new`.